### DISCUSSION TITLE
Interpretation of AspectRatioMode.FILL, ADJUST and NONE

### DISCUSSION THREAD

**Discussion by replicant1** [pedroSG94/RootEncoder#1546] [General]
Hi Pedro,

I'm having some trouble understanding what difference the three values of `AspectRatioMode` make to the preview image I see. After some experimentation, I see that `adjust` gives me a couple of black gutters at left and right in landscape orientation - but I can't tell if the image itself is slightly stretched horizontally or not. 

Does the `app:aspectRatioMode` value I set to an `OpenGlView` change the way the image is rendered, or does it only effect how much of the OpenGlView is devoted to the image? 

I hope you can see what I mean ... this is hard to put into words.  Maybe this image will make it clearer:

![where-did-it-go](https://github.com/user-attachments/assets/b5b40dec-d075-449d-b49c-612e7a2f0089)

This is my app in landscape mode. There's an `OpenGlView` in the background that occupies the full screen, and a layer of Jetpack Compose controls on top of that. The OpenGlView is in preview mode. Note the black gutters at left and right, marked up with white arrows. This comes from setting the `OpenGlView` 's `aspectRatioMode` to `adjust`. If I change the `aspectRatioMode` to "`fill`", those black gutters disappear, but I cannot tell if that results in the image being slightly stretched horizontally more or not. If that is happening, it might be too subtle to notice.

Can you please describe what each of the three `aspectRatioMode` values (`adjust`, `fill`, `none`) actually mean in terms of the rendered image? In particular, *what* is being adjusted if I choose `adjust`?  I've looked at the code in `SizeCalculator.getViewport` but couldn't understand it.

**In summary:** If it is important to me that the preview image should appear such that objects in the preview image appear with their natural proportions, which `aspectRatioMode`(s) should I use?

Thanks for your time.

Rod.

**Comment by pedroSG94** [comment]
Hello,

This is the difference between each value:
- Adjust -> Adapt image to full size of the preview view. If the preivew have a different aspect ratio then the preview is filled with black gutters to avoid distortion.
- Fill -> Adapt image to full size of the preview view. If the preivew have a different aspect ratio then the image cropped to avoid distortion.
- None -> Fill the image with the preview, doing distortion if necessary.

You have an image example here:
![scaleimage](https://github.com/user-attachments/assets/72b2d4b8-e758-440c-a4eb-37fbb3bb221d)

Where:
- Adjust -> centerInside (in this case black gutter are white).
- Fill -> centerCrop.
- None -> fitXY.

**Reply by replicant1** [reply]
Hi Pedro,

Thanks for your brilliant reply. Legend! You've cleared up my misunderstanding of `AspectRatioMode` completely!

 I wonder if I can stretch your patience a little further with a follow-up question?

You might remember in a previous question I sent you a link to a github repo that contains an app that demonstrates my question. I've updated that repo to illustrate a new question:

`https://github.com/replicant1/OpenGLViewRootEncoderProblem`

This one is very, very puzzling to me. If you run the app straight out of the repo you will see that it contains an `OpenGlView` in preview mode, and with the `AspectRatioMode` set to `adjust`. If you rotate your device a few times you will see that the `OpenGlView` is handling these orientation changes correctly. There are black gutters at the left and right of the image when in landscape orientation, just as you expect.

Here's where things get **freaky**.

If you open the file `InstructorStreamFragment.kt` and look in the `onCreateView()` method, you'll see two comments beginning `OPTION ONE` and `OPTION TWO`. These comments identify two different lines of code that, just to look at them, you would think would accomplish exactly the same thing - but they don't.

`OPTION ONE` is the following line: `myCameraHolder.save(binding.surfaceView, this)`
`OPTION TWO` is the following line: `myCameraHolder.save(RtmpCamera1(binding.surfaceView, this))`

`OPTION TWO` works in the expected manner., but `OPTION ONE` does not. With `OPTION ONE`, the resulting app does NOT handle device orientation changes in the same way. But the code surely looks like it should have exactly the same effect as `OPTION TWO` ... its just a matter of whether the `RtmpCamera1` instance is created inside `MyCameraHolder` or outside it. Why would that make any difference? I just can't see an explanation for it. 

It's an issue for me, because the MyCameraHolder is standing in for an Android ViewModel and in my real code base, I want to create the `RtmpCamera1` in a `ViewModel` so that it will persist across configuration changes. But as soon as I try this, the  `OPTION ONE `problem occurs.

It's always possible I'm missing something silly, but this one really has me stumped. I'm wondering if it doesn't expose some kind of memory issue. Anyway if you have time to check this out, I would appreciate it - but if it's too involved then I'll understand you giving it a miss.

Thanks for your responsiveness.

**Comment by pedroSG94** [comment]
Hello,

If you means the reason about aspect ratio is not working with the case 1 it is because you set a SurfaceView to RtmpCamera1 in the first case instead of an OpenGlView.
You need replace this:
https://github.com/replicant1/OpenGLViewRootEncoderProblem/blob/master/app/src/main/java/com/rodbailey/openglviewrootencoderproblem/MyCameraHolder.kt#L17
To:
```kotlin
    fun save(surfaceView: OpenGlView, connectChecker: ConnectChecker) {
```

The reason is because RtmpCamera1 accept both, OpenGlView and SurfaceView. In this case you are casting your OpenGlView to a SurfaceView and library can't know it so it is working as a SurfaceView ignoring aspect ratio (OpenGlView is a custom implementation of SurfaceView with cheats like the aspect ratio or filters)

**Reply by replicant1** [reply]
Hi Pedro,

Thanks for your quick response and diagnosis of the problem.
I feel so foolish - I have been looking at that for a long time and have just been looking straight past the issue. 

Thanks again!

Rod.
